File: eco-core/core-backend/src/main/java/com/willfp/eco/internal/config/EcoConfigFactory.java
package com.willfp.eco.internal.config;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.config.Config;
import com.willfp.eco.core.config.json.JSONConfig;
import com.willfp.eco.core.config.wrapper.ConfigFactory;
import com.willfp.eco.internal.config.json.JSONConfigSection;
import com.willfp.eco.internal.config.json.LoadableJSONConfig;
import com.willfp.eco.internal.config.yaml.LoadableYamlConfig;
import com.willfp.eco.internal.config.yaml.UpdatableYamlConfig;
import com.willfp.eco.internal.config.yaml.YamlConfigSection;
import org.bukkit.configuration.file.YamlConfiguration;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

public class EcoConfigFactory implements ConfigFactory {
    @Override
    public com.willfp.eco.core.config.yaml.LoadableYamlConfig createUpdatableYamlConfig(@NotNull final String configName,
                                                                                        @NotNull final EcoPlugin plugin,
                                                                                        @NotNull final String subDirectoryPath,
                                                                                        @NotNull final Class<?> source,
                                                                                        final boolean removeUnused,
                                                                                        @NotNull final String... updateBlacklist) {
        return new UpdatableYamlConfig(
                configName,
                plugin,
                subDirectoryPath,
                source,
                removeUnused,
                updateBlacklist
        );
    }

    @Override
    public com.willfp.eco.core.config.json.LoadableJSONConfig createLoadableJSONConfig(@NotNull final String configName,
                                                                                       @NotNull final EcoPlugin plugin,
                                                                                       @NotNull final String subDirectoryPath,
                                                                                       @NotNull final Class<?> source) {
        return new LoadableJSONConfig(
                configName,
                plugin,
                subDirectoryPath,
                source
        );
    }

    @Override
    public com.willfp.eco.core.config.yaml.LoadableYamlConfig createLoadableYamlConfig(@NotNull final String configName,
                                                                                       @NotNull final EcoPlugin plugin,
                                                                                       @NotNull final String subDirectoryPath,
                                                                                       @NotNull final Class<?> source) {
        return new LoadableYamlConfig(
                configName,
                plugin,
                subDirectoryPath,
                source
        );
    }

    @Override
    public Config createYamlConfig(@NotNull final YamlConfiguration config) {
        return new YamlConfigSection(config);
    }

    @Override
    public JSONConfig createJSONConfig(@NotNull final Map<String, Object> values) {
        return new JSONConfigSection(values);
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/config/EcoConfigFactory.kt
package com.willfp.eco.internal.config

import com.willfp.eco.core.PluginLike
import com.willfp.eco.core.config.ConfigType
import com.willfp.eco.core.config.interfaces.Config
import com.willfp.eco.core.config.interfaces.LoadableConfig
import com.willfp.eco.core.config.wrapper.ConfigFactory
import org.bukkit.configuration.ConfigurationSection

object EcoConfigFactory : ConfigFactory {
    override fun createConfig(bukkit: ConfigurationSection): Config {
        val config = createConfig(emptyMap(), ConfigType.YAML)
        for (key in bukkit.getKeys(true)) {
            config.set(key, bukkit.get(key))
        }

        return config
    }

    override fun createConfig(values: Map<String, Any>, type: ConfigType): Config =
        EcoConfigSection(type, values)

    override fun createConfig(contents: String, type: ConfigType): Config =
        EcoConfigSection(type, type.toMap(contents))

    override fun createLoadableConfig(
        configName: String,
        plugin: PluginLike,
        subDirectoryPath: String,
        source: Class<*>,
        type: ConfigType,
        requiresChangesToSave: Boolean
    ): LoadableConfig = EcoLoadableConfig(
        type,
        configName,
        plugin,
        subDirectoryPath,
        source,
        requiresChangesToSave
    )

    override fun createUpdatableConfig(
        configName: String,
        plugin: PluginLike,
        subDirectoryPath: String,
        source: Class<*>,
        removeUnused: Boolean,
        type: ConfigType,
        requiresChangesToSave: Boolean,
        vararg updateBlacklist: String
    ): LoadableConfig = EcoUpdatableConfig(
        type,
        configName,
        plugin,
        subDirectoryPath,
        source,
        removeUnused,
        requiresChangesToSave,
        *updateBlacklist
    )
}
